File: Haskell/countprimes2.hs
-- count_primes

sieve (p:xs) =  p : sieve [ x | x <- xs, x `mod` p > 0 ]
sieve ([])   =  []
        
main :: IO ()
main = do
    input <- getLine
    let number = read input :: Integer
    let primes = [ x | x <- sieve( [2..number] ) ]
    print $ length primes

-- Performance numbers:
--
-- 100000000
-- NOTE: The performance is rather bad, so I'm obviously doing something really wrong here.

File: Scala/countprimes2.scala
/* count_primes */

object Sieve {
    def sieve( xs: Stream[Int] ):Stream[Int] =
        xs match {
            case first #:: rest => first #:: sieve( rest.filter( x => x % first != 0 ) )
            case _ => Stream.empty
        }
 
    def main(args: Array[String]) {
        val number = readInt()
        val primes = sieve( List.range(2, number).toStream ).toList
        println( primes.length )
    }
}

/*
 * Performance numbers:
 *
 * 100000000 => ????
 */
